### ISSUE TITLE
/dev/shm woes on IDIA nodes

### ISSUE THREAD

**Issue by IanHeywood** [ratt-ru/CubiCal#311]
First this:

```
 - 02:11:24 - NpShared           [x07] [14.4/17.4 16.4/19.3 4.3Gb] File file:///dev/shm/cubical.16743/DATA:843696:876896/str:solutions:d/str:D 0T6F0:d/str:ifrgains:DDH__:a exists, deleting
```

then it fails with:

```
OSError: [Errno 2] No such file or directory: 'file:///dev/shm/cubical.16743/DATA:843696:876896/str:solutions:d/str:D0T6F0:d/str:ifrgains:DDH__:a'
```

More context:

```
- 02:04:23 - main               [io] [4.4/34.0 5.9/37.0 15.0Gb] I/O job(s) complete
 - 02:06:21 - solver             [x06] [56.1/69.1 58.2/72.1 14.8Gb] D0T5F0 (end solve) G: 20 iters, conv 87.50%, d/fl 26.21%, PGE 0.0234, stall 96.29
%, chi^2 1.299 -> 1.209 (1.525), noise 0.247 -> 0.217
 - 02:07:42 - parallactic_machin [x06] [65.2/73.3 67.2/75.4 19.3Gb] Applying P Jones to corrected data (precomputed)
 - 02:08:34 - main               [1.0/2.0 3.3/3.7 19.3Gb] finished processing tile 5/49
 - 02:08:34 - main               [1.0/2.0 3.3/3.7 19.3Gb] waiting for I/O on tile 6/49
 - 02:08:34 - main               [1.0/2.0 3.3/3.7 19.3Gb] submitting solver jobs for tile 6/49
 - 02:08:34 - main               [io] [4.3/34.0 5.7/37.0 19.3Gb] saving tile 5/49
 - 02:08:34 - data_handler       [io] [4.3/34.0 5.7/37.0 19.3Gb] tile 5/49: saving MS rows 703080~843695
 - 02:08:34 - data_handler       [io] [4.3/34.0 10.0/37.0 19.3Gb]   writing CORRECTED_DATA column
 - 02:08:41 - parallactic_machin [x07] [6.8/7.8 8.7/9.7 19.3Gb] Applying P Jones to sky (precomputed)
 - 02:11:14 - solver             [x07] [14.4/17.4 16.4/19.3 4.3Gb] D0T6F0 (end solve) G: 20 iters, conv 87.50%, d/fl 26.21%, PGE 0.0639, stall 95.98%
, chi^2 1.286 -> 1.199 (1.508), noise 0.249 -> 0.219
 - 02:11:24 - NpShared           [x07] [14.4/17.4 16.4/19.3 4.3Gb] File file:///dev/shm/cubical.16743/DATA:843696:876896/str:solutions:d/str:D
0T6F0:d/str:ifrgains:DDH__:a exists, deleting
 - 02:11:24 - solver             [x07] [14.4/17.4 16.4/19.3 4.3Gb] Solver for tile 6 chunk D0T6F0 failed with exception: [Errno 2] No such file or di
rectory: 'file:///dev/shm/cubical.16743/DATA:843696:876896/str:solutions:d/str:D0T6F0:d/str:ifrgains:DDH__:a'
 - 02:11:24 - solver             [x07] [14.4/17.4 16.4/19.3 4.3Gb] Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/cubical/solver.py", line 833, in run_solver
    corr_vis = solver_machine.run()
  File "/usr/local/lib/python2.7/dist-packages/cubical/solver.py", line 656, in run
    SolveOnly.run(self)
  File "/usr/local/lib/python2.7/dist-packages/cubical/solver.py", line 644, in run
    self.vdm.freq_slice, self.soldict)
  File "/usr/local/lib/python2.7/dist-packages/cubical/machines/ifr_gain_machine.py", line 123, in update
    DDH = soldict.addSharedArray('ifrgains:DDH__', bbshape, dtype=obser_arr.dtype)
  File "/usr/local/lib/python2.7/dist-packages/cubical/tools/shared_dict.py", line 285, in addSharedArray
    array = NpShared.CreateShared(_to_shm(filepath), shape, dtype)
  File "/usr/local/lib/python2.7/dist-packages/cubical/tools/NpShared.py", line 45, in CreateShared
    a = SharedArray.create(Name, shape, dtype=dtype)
OSError: [Errno 2] No such file or directory: 'file:///dev/shm/cubical.16743/DATA:843696:876896/str:solutions:d/str:D0T6F0:d/str:ifrgains:DDH__:a'
 - 02:18:12 - data_handler       [io] [8.6/34.0 10.0/37.0 4.3Gb]   no new flags were generated
 - 02:18:12 - main               [io] [2.1/34.0 3.5/37.0 0.0Gb] I/O handler for load 7 save 5 failed with exception: [Errno 2] No such file or
 directory: '/dev/shm/cubical.16743/DATA:703080:843695'
 - 02:18:13 - main               [io] [2.1/34.0 3.5/37.0 0.0Gb] Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/cubical/workers.py", line 437, in _io_handler
    for sd in tile.iterate_solution_chunks():
  File "/usr/local/lib/python2.7/dist-packages/cubical/data_handler/ms_tile.py", line 1118, in iterate_solution_chunks
    data = shared_dict.attach(self._data_dict_name)
  File "/usr/local/lib/python2.7/dist-packages/cubical/tools/shared_dict.py", line 28, in attach
    return SharedDict(name, reset=False, load=load, readwrite=readwrite)
  File "/usr/local/lib/python2.7/dist-packages/cubical/tools/shared_dict.py", line 104, in __init__
    self.delete()
  File "/usr/local/lib/python2.7/dist-packages/cubical/tools/shared_dict.py", line 136, in delete
    os.mkdir(self.path)
OSError: [Errno 2] No such file or directory: '/dev/shm/cubical.16743/DATA:703080:843695'
 - 02:18:14 - main               [1.0/2.0 3.2/3.7 0.0Gb] Exiting with exception: OSError([Errno 2] No such file or directory: 'file:///dev/shm/cubical.1674
3/DATA:843696:876896/str:solutions:d/str:D0T6F0:d/str:ifrgains:DDH__:a')
 Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/cubical/main.py", line 480, in main
    stats_dict = workers.run_process_loop(ms, tile_list, load_model, single_chunk, solver_type, solver_opts, debug_opts)
  File "/usr/local/lib/python2.7/dist-packages/cubical/workers.py", line 210, in run_process_loop
    return _run_multi_process_loop(ms, load_model, solver_type, solver_opts, debug_opts)
  File "/usr/local/lib/python2.7/dist-packages/cubical/workers.py", line 296, in _run_multi_process_loop
    stats = future.result()
  File "/usr/local/lib/python2.7/dist-packages/concurrent/futures/_base.py", line 455, in result
    return self.__get_result()
  File "/usr/local/lib/python2.7/dist-packages/concurrent/futures/_base.py", line 414, in __get_result
    raise exception_type, self._exception, self._traceback
OSError: [Errno 2] No such file or directory: 'file:///dev/shm/cubical.16743/DATA:843696:876896/str:solutions:d/str:D0T6F0:d/str:ifrgains:DDH__:a'
```

**Comment by IanHeywood** [comment]
The usual for an IDIA slurm worker node:

```
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           119G     0  119G   0% /dev/shm
```

**Comment by o-smirnov** [comment]
Have they applied this to the nodes? Can't remember if we discussed it... https://gitlab.com/tenzing/shared-array#on-linux-i-get-cannot-allocate-memory-when-creating-many-arrays

**Comment by IanHeywood** [comment]
There's presently one worker node that has `vm.max_map_count = 1000000` and `192G   0% /dev/shm`. I'm running it on there now, but it's taking its sweet time. This is all running with `ncpu = 0`, just trying to get the bastard to actually finish.

**Comment by IanHeywood** [comment]
Same outcome.

**Comment by IanHeywood** [comment]
Actually `vm.max_map_count = 65530`, but /dev/shm is 192G.

**Comment by JSKenyon** [comment]
I don't have much insight into the problem, but I think that `vm.max_map_count` likely needs to be set higher than that. A common number seems to be 262144. Stills seems to have it set to 100000.

**Comment by bennahugo** [comment]
It is either raising the counts, or try enabling hugepages on the machine. The logic behind that would be if you have large contiguous blocks of memory you will need fewer pages. It may also improve the performance of the translation lookaside buffer (which has limited slots available).

**Comment by bennahugo** [comment]
For cubical as far as I know we mostly only store very large chunks of memory in the cache (apart from perhaps a few things that probably should not live in shared memory in the first place)

**Comment by bennahugo** [comment]
Ask your friendly IDIA dev ops to look into
http://manpages.ubuntu.com/manpages/trusty/man8/hugeadm.8.html

And give him/her the argument along the lines:
For me I have 4 MB (```getconf PAGE_SIZE```) pages on my machine. If  we assume a 100GB of visibilities alone we are already sitting on 25600 page entries already just for visibilities alone!

**Comment by o-smirnov** [comment]
Yeah I agree, either get them to up vm.max_map_count, or enable huge pages, or ideally both. I'll push a change to NpShared to make the error message more accurate.

I don't wholly understand the details, but e.g. these guys: https://docs.actian.com/vector/5.0/index.html#page/User/Increase_max_map_count_Kernel_Parameter_(Linux).htm

...recommend 1 map count per 128KB of memory. Which works out to 2 million for a 256GB node.

Basically the bottom line is, if you take a big memory node and install a stock Linux on it, it's not really configured to let applications take advantage of that memory out of the box. Admins need to do this sort of tuning.

**Comment by bennahugo** [comment]
For many small arrays smaller page sizes are good otherwise large pages are huge. I would argue the other way around for radio astronomy. Specifically to optimize the TLB performance. I've shown as much for some scaling figures I worked out a while back

**Comment by bennahugo** [comment]
![image](https://user-images.githubusercontent.com/7934586/63931937-86dadf80-ca56-11e9-9152-6b02fe3e217d.png)
right issue this time...

**Comment by bennahugo** [comment]
For larger arrays the allocation and memsetting time improves substantially by not thrashing the TLB with a multitude of small pages

**Comment by IanHeywood** [comment]
I'm getting the same outcome as the original post with `sysctl -w vm.max_map_count=1000000` on the 192 /dev/shm node. This is a conservative run with parameters that were completing just fine (albeit after six hours) on stock nodes back in January, so I don't know WTF is going on.

**Comment by IanHeywood** [comment]
Enabled 8 and 16 rebinning in time and frequency respectively, and it got as far as tile 8/14 before crashing out with the usual:

```
OSError: [Errno 2] No such file or directory: 'file:///dev/shm/cubical.5493/DATA:3511494:4011461/str:covis:a'
```

I checked in on the node shortly beforehand during a solve and `htop` said I was using about a quarter of the available RAM, and it didn't look like /dev/shm was exactly being stretched:

```
ianh@slwrk-117:~$ df -h /dev/shm/
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           119G  222M  118G   1% /dev/shm
```

**Comment by o-smirnov** [comment]
Weird weird weird. @cyriltasse ring any bells with your cluster woes?

I'll make a branch with more detailed error reporting that we can try. @IanHeywood, is it possible at all to move the MS to nash and give it a go there? Just for comparative purposes?

**Comment by IanHeywood** [comment]
> Also which version is this? Is the bahaviour the same in the latest release?

I'm using @SpheMakh's stimela container for cubical, 1.1.5. The test I'm trying now is to go back to the (much more fiddly) venv setup.

PS: Sorry this issue is basically turning into my blog.

**Comment by IanHeywood** [comment]
> I'll make a branch with more detailed error reporting that we can try. @IanHeywood, is it possible at all to move the MS to nash and give it a go there? Just for comparative purposes?

This is running as we speak, same container, using the parset with no rebinning. Currently on tile 14/48 after three hours. (Like I said, it's a conservative setup. Using nash this way must be how you feel when you're sitting in traffic in your Golf.)

```
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           454G   34G  420G   8% /dev/shm
```

**Comment by IanHeywood** [comment]
> It is odd that it runs to about 8 tiles. It may indicate a steady accumulation in shm pages. Are we sure memory is released properly?

Here's my totally rigorous examination of this possibility on nash, using 'tail' and 'df'. At least there it looks OK.

```
 - 11:04:22 - data_handler       [9.2/38.4 9.7/39.0 14.2Gb] tile 16/48: reading MS rows 1807796~1928819
 - 11:04:22 - data_handler       [9.2/38.4 9.7/39.0 14.2Gb] reading DATA
 - 11:04:33 - data_handler       [12.9/38.4 15.2/39.0 14.2Gb] model 0 weights 0: reading from WEIGHT_SPECTRUM
 - 11:04:38 - data_handler       [16.6/38.4 18.0/39.0 14.2Gb]   applying solvable baseline cutoff deselects 9358 rows
 - 11:04:38 - data_handler       [16.6/38.4 18.0/39.0 14.2Gb]   7.73% visibilities deselected via specificed subset and/or baseline cutoffs
 - 11:04:38 - data_handler       [16.6/38.4 18.0/39.0 14.2Gb] reading BITFLAG
 - 11:05:00 - data_handler       [20.3/38.4 20.8/39.0 14.2Gb]   72.24% input visibilities flagged and/or deselected
 - 11:05:06 - data_handler       [24.9/38.4 29.1/39.0 20.7Gb]   reading MODEL_DATA for model 0 direction 0
 - 11:05:06 - data_handler       [24.9/38.4 29.1/39.0 20.7Gb] reading MODEL_DATA
 - 11:11:16 - solver             [27.4/38.4 27.9/39.0 24.4Gb] D0T32F0 (end solve) G: 20 iters, conv 75.00%, d/fl 16.39%, PGE 0.051, stall 97.30%, chi^2 0.3937 -> 0.3921 (0.4265), noise 0.228 -> 0.218

Filesystem      Size  Used Avail Use% Mounted on
tmpfs           454G   38G  416G   9% /dev/shm

 - 11:29:38 - data_handler       [7.9/38.4 8.5/39.0 14.2Gb] tile 18/48: reading MS rows 2029043~2150066
 - 11:29:38 - data_handler       [7.9/38.4 8.5/39.0 14.2Gb] reading DATA
 - 11:29:50 - data_handler       [11.6/38.4 14.0/39.0 14.2Gb] model 0 weights 0: reading from WEIGHT_SPECTRUM
 - 11:29:55 - data_handler       [15.3/38.4 16.8/39.0 14.2Gb]   applying solvable baseline cutoff deselects 8960 rows
 - 11:29:55 - data_handler       [15.3/38.4 16.8/39.0 14.2Gb]   7.40% visibilities deselected via specificed subset and/or baseline cutoffs
 - 11:29:55 - data_handler       [15.4/38.4 16.8/39.0 14.2Gb] reading BITFLAG
 - 11:30:18 - data_handler       [19.0/38.4 19.5/39.0 14.2Gb]   71.88% input visibilities flagged and/or deselected
 - 11:30:23 - data_handler       [23.6/38.4 27.8/39.0 20.7Gb]   reading MODEL_DATA for model 0 direction 0
 - 11:30:23 - data_handler       [23.6/38.4 27.8/39.0 20.7Gb] reading MODEL_DATA
 - 11:36:36 - solver             [27.4/38.4 27.9/39.0 24.4Gb] D0T36F0 (end solve) G: 20 iters, conv 75.00%, d/fl 16.39%, PGE 0.0479, stall 97.28%, chi^2 0.9928 -> 0.989 (1.079), noise 0.224 -> 0.213

Filesystem      Size  Used Avail Use% Mounted on
tmpfs           454G   38G  416G   9% /dev/shm

 - 13:37:28 - data_handler       [7.9/38.4 8.5/39.0 14.2Gb] tile 28/48: reading MS rows 3148515~3269538
 - 13:37:28 - data_handler       [7.9/38.4 8.5/39.0 14.2Gb] reading DATA
 - 13:37:40 - data_handler       [11.6/38.4 14.0/39.0 14.2Gb] model 0 weights 0: reading from WEIGHT_SPECTRUM
 - 13:37:45 - data_handler       [15.3/38.4 16.8/39.0 14.2Gb]   applying solvable baseline cutoff deselects 8384 rows
 - 13:37:45 - data_handler       [15.3/38.4 16.8/39.0 14.2Gb]   6.93% visibilities deselected via specificed subset and/or baseline cutoffs
 - 13:37:45 - data_handler       [15.4/38.4 16.8/39.0 14.2Gb] reading BITFLAG
 - 13:38:12 - data_handler       [19.0/38.4 19.5/39.0 14.2Gb]   71.62% input visibilities flagged and/or deselected
 - 13:38:18 - data_handler       [23.6/38.4 27.8/39.0 20.7Gb]   reading MODEL_DATA for model 0 direction 0
 - 13:38:18 - data_handler       [23.6/38.4 27.8/39.0 20.7Gb] reading MODEL_DATA
 - 13:44:27 - solver             [27.4/38.4 27.9/39.0 24.4Gb] D0T56F0 (end solve) G: 20 iters, conv 75.00%, d/fl 16.39%, PGE 0.0524, stall 97.12%, chi^2 0.9839 -> 0.9814 (1.074), noise 0.228 -> 0.217
ianh@nash:~/LADUMA$ df -h /dev/shm
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           454G   38G  416G   9% /dev/shm
```

**Comment by IanHeywood** [comment]
Version 1.3.0 inside a venv has the same outcome:

```
INFO       - 15:53:55 - data_handler       [0.7/2.1 2.9/4.3 9.1Gb] WARNING: output columns will be upsampled from time-binned data!
INFO       - 15:53:56 - data_handler       [0.8/2.1 3.7/4.3 9.1Gb]   generated 105 row chunks based on time and DDID
INFO       - 15:53:56 - data_handler       [0.8/2.1 3.7/4.3 9.1Gb]   row chunks yield 105 potential tiles
INFO       - 15:53:56 - data_handler       [0.9/2.1 3.8/4.3 9.1Gb]   coarsening this to 53 tiles (max 2 chunks per tile, based on 1/0 requested)
INFO       - 15:54:04 - data_handler       [0.7/2.1 3.6/4.3 9.1Gb] the MS appears to have a properly formed BITFLAG column
INFO       - 15:54:04 - data_handler       [0.7/2.1 3.6/4.3 9.1Gb]   bitflag 'legacy' already exists, will not auto-fill
INFO       - 15:54:04 - data_handler       [0.7/2.1 3.6/4.3 9.1Gb]   BITFLAG column defines the following flagsets: legacy:1 cubical:2
INFO       - 15:54:04 - data_handler       [0.7/2.1 3.6/4.3 9.1Gb]   will exclude flagset cubical
INFO       - 15:54:04 - data_handler       [0.7/2.1 3.6/4.3 9.1Gb]   applying BITFLAG mask 1 to input data
INFO       - 15:54:04 - data_handler       [0.7/2.1 3.6/4.3 9.1Gb]   will save output flags into BITFLAG 'cubical' (2), and into FLAG/FLAG_RO
W
INFO       - 15:54:04 - ms_tile            [0.7/2.1 3.6/4.3 9.1Gb] tile 0/53: reading MS rows 0~124991
INFO       - 15:54:04 - data_handler       [0.7/2.1 3.6/4.3 9.1Gb] reading DATA
INFO       - 15:58:25 - data_handler       [8.3/8.3 12.2/13.1 9.1Gb] reading BITFLAG
INFO       - 16:01:26 - ms_tile            [12.6/14.5 15.0/16.9 9.1Gb]   69.24% input visibilities flagged and/or deselected
INFO       - 16:01:26 - ms_tile            [12.6/14.5 15.0/16.9 9.1Gb]   rebinning into 15624 rows and 64 channels
2019-08-30 16:01:31	WARN	MeasIERS::findTab (file /build/casacore-IDmUo1/casacore-2.4.1/measures/Measures/MeasIERS.cc, line 381)	Requested data table IER
Seop97 cannot be found in the searched directories:
2019-08-30 16:01:31	WARN	MeasIERS::findTab (file /build/casacore-IDmUo1/casacore-2.4.1/measures/Measures/MeasIERS.cc, line 381)+	/var/lib/casacore/data/e
phemerides/
2019-08-30 16:01:31	WARN	MeasIERS::findTab (file /build/casacore-IDmUo1/casacore-2.4.1/measures/Measures/MeasIERS.cc, line 381)+	/var/lib/casacore/data/g
eodetic/
2019-08-30 16:01:31	WARN	MeasIERS::findTab (file /build/casacore-IDmUo1/casacore-2.4.1/measures/Measures/MeasIERS.cc, line 381)	Requested data table IER
Spredict cannot be found in the searched directories:
2019-08-30 16:01:31	WARN	MeasIERS::findTab (file /build/casacore-IDmUo1/casacore-2.4.1/measures/Measures/MeasIERS.cc, line 381)+	/var/lib/casacore/data/e
phemerides/
2019-08-30 16:01:31	WARN	MeasIERS::findTab (file /build/casacore-IDmUo1/casacore-2.4.1/measures/Measures/MeasIERS.cc, line 381)+	/var/lib/casacore/data/g
eodetic/
2019-08-30 16:01:31	INFO	MeasIERS::fillMeas(MeasIERS::Files, Double) 	Requested JD 58657.2 is outside the range of the IERS (Earth axis data) table.
2019-08-30 16:01:31	INFO	MeasIERS::fillMeas(MeasIERS::Files, Double) +	Calculations will proceed with less precision
2019-08-30 16:01:31	SEVERE	MeasTable::dUTC(Double) (file /build/casacore-IDmUo1/casacore-2.4.1/measures/Measures/MeasTable.cc, line 4396)	Leap second tabl
e TAI_UTC seems out-of-date.
2019-08-30 16:01:31	SEVERE	MeasTable::dUTC(Double) (file /build/casacore-IDmUo1/casacore-2.4.1/measures/Measures/MeasTable.cc, line 4396)+	Until the table
is updated (see the CASA documentation or your system admin),
2019-08-30 16:01:31	SEVERE	MeasTable::dUTC(Double) (file /build/casacore-IDmUo1/casacore-2.4.1/measures/Measures/MeasTable.cc, line 4396)+	times and coordi
nates derived from UTC could be wrong by 1s or more.
INFO       - 16:01:31 - data_handler       [8.8/14.5 11.3/16.9 9.2Gb] reading MODEL_DATA
INFO       - 16:11:51 - NpShared           [8.0/16.5 10.3/18.9 9.2Gb] File file:///dev/shm/cubical.3230/DATA:0:124991/str:covis:a exists, del
eting
INFO       - 16:11:51 - main               [8.0/16.5 10.3/18.9 9.2Gb] Exiting with exception: OSError([Errno 2] No such file or directory: 'f
ile:///dev/shm/cubical.3230/DATA:0:124991/str:covis:a')
 Traceback (most recent call last):
  File "/users/ianh/Software/CubiCal/cubical/main.py", line 534, in main
    stats_dict = workers.run_process_loop(ms, tile_list, load_model, single_chunk, solver_type, solver_opts, debug_opts, out_opts)
  File "/users/ianh/Software/CubiCal/cubical/workers.py", line 216, in run_process_loop
    return _run_single_process_loop(ms, load_model, single_chunk, solver_type, solver_opts, debug_opts, out_opts)
  File "/users/ianh/Software/CubiCal/cubical/workers.py", line 347, in _run_single_process_loop
    tile.load(load_model=load_model)
  File "/users/ianh/Software/CubiCal/cubical/data_handler/ms_tile.py", line 1022, in load
    data.addSharedArray('covis', data['obvis'].shape, self.dh.ctype)
  File "/users/ianh/Software/CubiCal/cubical/tools/shared_dict.py", line 285, in addSharedArray
    array = NpShared.CreateShared(_to_shm(filepath), shape, dtype)
  File "/users/ianh/Software/CubiCal/cubical/tools/NpShared.py", line 45, in CreateShared
    a = SharedArray.create(Name, shape, dtype=dtype)
OSError: [Errno 2] No such file or directory: 'file:///dev/shm/cubical.3230/DATA:0:124991/str:covis:a'
```

**Comment by IanHeywood** [comment]
How do CubiCal runs invoked by MeerKATHI behave on IDIA nodes?

**Comment by IanHeywood** [comment]
Small tiles, rebinning, and a serial run failing on worker node with `vm.max_map_count=262144` and 220 GB of /dev/shm...

```
  File "/users/ianh/Software/CubiCal/cubical/workers.py", line 347, in _run_single_process_loop
    tile.load(load_model=load_model)
  File "/users/ianh/Software/CubiCal/cubical/data_handler/ms_tile.py", line 872, in load
    obvis = data.addSharedArray('obvis', [nrows, nchan, self.dh.ncorr], obvis0.dtype)
  File "/users/ianh/Software/CubiCal/cubical/tools/shared_dict.py", line 285, in addSharedArray
    array = NpShared.CreateShared(_to_shm(filepath), shape, dtype)
  File "/users/ianh/Software/CubiCal/cubical/tools/NpShared.py", line 45, in CreateShared
    a = SharedArray.create(Name, shape, dtype=dtype)
OSError: [Errno 2] No such file or directory: 'file:///dev/shm/cubical.26026/DATA:0:124991/str:obvis:a'
```

There's no way this thing is actually running out of memory, surely...

**Comment by IanHeywood** [comment]
Meanwhile my initial test with Sphe's stimela image on nash has strolled over the finish line, so it's not the container that's the problem.

```
Time taken for SolveAndCorrect: 34848.8998048 seconds
```

```
[3.1/38.4 3.4/39.0 14.0Gb]
```

**Comment by JSKenyon** [comment]
I am not 100% up to date but I think the issue may be that the pypi version is massively out of date. I would also be suspicious of your virtual environment - tear it down and install directly from the master branch on git. 

Something like ‘pip install cubical@git+https://github.com/ratt-ru/CubiCal.git’ should sort it. 

https://github.com/ratt-ru/CubiCal.git

**Comment by IanHeywood** [comment]
Thanks @JSKenyon. 

If I'm referring to `v1.3.0` then that's the current master branch inside a virtual environment that I installed yesterday. On IDIA I have to start a singularity shell inside a kern4 container to set this up, following the same procedure as back when this all used to work.

Could something inside the container have gone stale to cause this? I'm not even sure where to look to check what you think might be the issue. Suggestions welcome.

An overnight run on the 262144 / 220 node is making better progress than any other test, but everything is dialled back to give an estimated completion time of about 32 hours so it's kind of pointless.

Cheers.

**Comment by IanHeywood** [comment]
> An overnight run on the 262144 / 220 node is making better progress than any other test, but everything is dialled back to give an estimated completion time of about 32 hours so it's kind of pointless.

```
  File "/users/ianh/Software/CubiCal/cubical/workers.py", line 347, in _run_single_process_loop
    tile.load(load_model=load_model)
  File "/users/ianh/Software/CubiCal/cubical/data_handler/ms_tile.py", line 1022, in load
    data.addSharedArray('covis', data['obvis'].shape, self.dh.ctype)
  File "/users/ianh/Software/CubiCal/cubical/tools/shared_dict.py", line 285, in addSharedArray
    array = NpShared.CreateShared(_to_shm(filepath), shape, dtype)
  File "/users/ianh/Software/CubiCal/cubical/tools/NpShared.py", line 45, in CreateShared
    a = SharedArray.create(Name, shape, dtype=dtype)
OSError: [Errno 2] No such file or directory: 'file:///dev/shm/cubical.26300/DATA:2757636:2820131/str:covis:a'
```

It's a good job I'm not *too* paranoid.

**Comment by o-smirnov** [comment]
OK, well if it runs on nash then there's nothing wrong with the machinery per se. There may be something system-specific about how limits on shared memory and things like ``vm.max_map_count`` are propagated into containers. My hypothesis is that it's some configuration setting on the IDIA nodes that hobbles it, but I don't understand enough about Singularity to know. @SpheMakh @bennahugo @gijzelaerr any insights?

**Comment by IanHeywood** [comment]
Right, even the virtualenv current master version runs inside a container...

Cheers.

**Comment by gijzelaerr** [comment]
no idea. A contained process is just a process. Maybe an IDIA specific restriction on processes?

**Comment by IanHeywood** [comment]
Successful run on a standalone IDIA node (i.e. not a slurm worker). Slow as hell again, but at least it completed, and I'm not sure what competition for resources on the standalone node was like during the run (and there is no vigilant slurm controller to kill the job the moment it steps out of line).

```
Time taken for SolveOnly: 61979.462219 seconds
```

I think at this stage everything points to a configuration issue with the slurm workers, rather than a CubiCal bug, a container issue, or me screwing up. I'll try to distill this into a support ticket for the IDIA folks.

Cheers.

**Comment by SpheMakh** [comment]
Hey Ian, singularity does not have an option to set shared-memory, and I think the default is very low because I've had this issue before on a 6GB data set. The good news is that you can set shared memory using podman which is basically a far advanced version of singularity (as advanced as docker but does not require root privileges to run containers). 

See the podman repo [here](https://github.com/containers/libpod)

**Comment by SpheMakh** [comment]
stimela also supports podman, or you can pull the stimela images using podman as you have been doing with singularity.

**Comment by bennahugo** [comment]
In principle there is no difference between the container and its host in terms of the page limits, but you are possibly running into other instances since you are on a shared pool resource. 
You should try running singularity in its own IPC namespace to ensure that there isn't any interplay with the system /dev/shm with the --ipc switch.  

Viewing this inside and outside should also provide some answers:
```
bhugo@bhugo-wks ~                                                                                   [16:35:42] 
> $ sysctl vm.max_map_count                                                                                   
vm.max_map_count = 65530
                                                                                                               
bhugo@bhugo-wks ~                                                                                   [16:36:16] 
> $ ipcs -lm                                                                                                  

------ Shared Memory Limits --------
max number of segments = 4096
max seg size (kbytes) = 18014398509465599
max total shared memory (kbytes) = 18014398509481980
min seg size (bytes) = 1
```

**Comment by IanHeywood** [comment]
Thanks @SpheMakh. If that is the cause then it's at odds with the fact that CubiCal was working, definitely in January and probably more recently, using singularity containers (kern containers running a virtualenv) on worker nodes with default /dev/shm settings.

**Comment by IanHeywood** [comment]
>In principle there is no difference between the container and its host in terms of the page limits, but you are possibly running into other instances since you are on a shared pool resource.

On the worker nodes I generally shouldn't be sharing resources with other users. I say generally as the other day someone had directly logged into a worker node that I was assigned and was running a bunch of resource-heavy stuff.

**Comment by bennahugo** [comment]
```
ipcs --human 
```
should shed some light on what and who is using shared array

**Comment by bennahugo** [comment]
It could be that the resources were never released?

**Comment by bennahugo** [comment]
but try the --ipc (http://singularity.lbl.gov/archive/docs/v2-3/docs-exec) call it should work around this issue

**Comment by SpheMakh** [comment]
or try the  `--containall` option to replicate the host settings on the container, but if the shared memory set low on the host, then there is nothing you can do.

**Comment by IanHeywood** [comment]
A possible clue here that suggests the `containall` suggestion might have mileage. I'm running @bennahugo's DDFacet container on nash, which throws up this warning:

```
 - 11:34:58 - DDFacet            | WARNING: max shared memory size is only 50% of total RAM size.
            This can cause problems for large imaging jobs. A setting of 90% is recommended for
            DDFacet and killMS. If your processes keep failing with SIGBUS or "bus error" messages,
            it is most likely for this reason. You can change the memory size by running
                $ sudo mount -o remount,size=90% /dev/shm
            To make the change permanent, edit /etc/defaults/tmps, and add a line saying "SHM_SIZE=90%".
```

However, outside of the container:

```
ianh@nash:~$ df -h /dev/shm/
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           454G  141G  313G  31% /dev/shm
```

which is 90% of the total RAM by my reckoning.

**Comment by IanHeywood** [comment]
Part of this problem might be that sometimes CubiCal (and DDFacet) do not clean up after themselves. I just got an out of disk space error running DDFacet on slwrk-155 (which is the 262144 / 220 test node, which is an unusual one, but I guess it was referring to /dev/shm as the /ceph area still has plenty remaining.

With (ostensibly) nothing running on the node I logged in and:

```
ianh@slwrk-155:~$ df -h /dev/shm/
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           220G  167G   54G  76% /dev/shm
```

manually running `CleanSHM.py` fixed most of this:

```
ianh@slwrk-155:~$ singularity shell ~/containers/ddfacet-0.4.1.simg
Singularity: Invoking an interactive shell within container...

Singularity ddfacet-0.4.1.simg:~> CleanSHM.py
 - 08:17:11 - ClearSHM           | Clear shared memory
 - 08:17:11 - Multiprocessing    | reaping 3374 shared memory objects associated with 1 dead DDFacet processes
 - 08:17:23 - ClearSHM           | Clear Semaphores
 - 08:17:23 - ClearSHM           | Clear shared dictionaries
```

but not all of it:

```
Singularity ddfacet-0.4.1.simg:~> df -h /dev/shm/
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           220G   24G  197G  11% /dev/shm
```

`htop` revealed some hung CubiCal processes, even though I had (so I thought) nothing running on this node, an assumption the slurm queue agreed with.

```
ianh@slwrk-155:~$ killall gocubical
ianh@slwrk-155:~$ htop
ianh@slwrk-155:~$ df -h /dev/shm/
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           220G   15G  206G   7% /dev/shm
```

Getting there...

```
ianh@slwrk-155:~$ singularity shell ~/containers/cubical-1.1.5.simg
Singularity: Invoking an interactive shell within container...

Singularity cubical-1.1.5.simg:~> gocubical -h
```

And finally.

```
Singularity cubical-1.1.5.simg:~> df -h /dev/shm
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           220G     0  220G   0% /dev/shm
```

**Comment by JSKenyon** [comment]
The combination of multiprocessing and shared memory in conjunction with interrupted/killed runs makes cleaning up a relatively difficult problem. I am currently hard at work on version 2 which makes use of threads and Dask. It is still a little way off, but I can already confirm that it dies cleanly when interrupted. @IanHeywood Your woes are motivating me to work faster.

**Comment by JSKenyon** [comment]
Close was clicked in error.

**Comment by IanHeywood** [comment]
Not sure that `--containall` is what we're after:

```
–contain: Contain suggests that we want to better isolate the container runtime from the host. Adding the --contain flag will use minimal /dev and empty other directories (e.g., /tmp).
–containall: In addition to what is provided with --contain (filesystems) also contain PID, IPC, and environment.
```

Indeed when I tried it I couldn't see the `/scratch/` or `/ceph` filesystems.
